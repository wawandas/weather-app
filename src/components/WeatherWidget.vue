<template>
  <v-card
    class="mx-auto"
    max-width="400"
    :loading="isLoading"
    :disabled="isLoading"
  >
    <v-form class="pa-4">
      <v-row>
        <v-col cols="10">
          <v-text-field
            name="city"
            placeholder="Enter city"
            solo
            clearable
            hide-details
            v-model="city"
            @keydown.enter.prevent="getWeatherForCity"
          >
          </v-text-field>
        </v-col>
        <v-col cols="2" class="align-center">
          <v-btn icon @click.prevent="getWeatherForCoords">
            <v-icon>mdi-map-marker</v-icon>
          </v-btn>
        </v-col>
      </v-row>
    </v-form>
    <div v-if="data.cod === 200">
      <v-list-item two-line>
        <v-list-item-content>
          <v-list-item-title class="headline" data-label="weather-city-name">{{
            data.name
          }}</v-list-item-title>
          <v-list-item-subtitle>
            {{ data.dt | formatedDate }}, {{ data.weather[0].main }}
          </v-list-item-subtitle>
        </v-list-item-content>
      </v-list-item>

      <v-card-text>
        <v-row align="center">
          <v-col cols="6">
            <span class="display-3">{{ data.main.temp | celsius }}</span>
          </v-col>
          <v-col cols="6" class="justify-center">
            <img :src="getImage()" />
            <br />
            ({{ data.weather[0].description }})
          </v-col>
        </v-row>
      </v-card-text>

      <v-list-item>
        <v-list-item-icon>
          <v-icon>mdi-send</v-icon>
        </v-list-item-icon>
        <v-list-item-subtitle>{{ data.wind.speed }} m/c</v-list-item-subtitle>
      </v-list-item>

      <v-list-item>
        <v-list-item-icon>
          <v-icon>mdi-cloud-download</v-icon>
        </v-list-item-icon>
        <v-list-item-subtitle>{{ data.main.humidity }}%</v-list-item-subtitle>
      </v-list-item>

      <v-card-actions>
        <v-btn icon>
          <v-icon>mdi-cached</v-icon>
        </v-btn>
      </v-card-actions>
    </div>
  </v-card>
</template>

<script>
export default {
  data: () => ({
    city: "",
    latitude: "",
    longitude: ""
  }),

  beforeMount() {
    this.getLocation();
  },

  computed: {
    isLoading() {
      return this.$store.state.isLoading;
    },

    data() {
      return this.$store.state.data;
    }
  },

  methods: {
    getLocation() {
      if (navigator.geolocation) {
        navigator.geolocation.getCurrentPosition(this.setCoords);
      } else {
        //x.innerHTML = "Geolocation is not supported by this browser.";
      }
    },

    setCoords(position) {
      this.longitude = position.coords.longitude;
      this.latitude = position.coords.latitude;
    },

    getWeatherForCity() {
      const query = `?q=${this.city}`;
      this.fetchData(query);
    },

    getWeatherForCoords() {
      this.getLocation();
      const query = `?lat=${this.latitude}&lon=${this.longitude}`;
      this.fetchData(query);
    },

    fetchData(query) {
      this.$store.dispatch("fetchData", query);
      this.city = "";
    },

    getImage() {
      return require(`../assets/weatherIcons/${this.data.weather[0].icon}.png`);
    }
  }
};
</script>
