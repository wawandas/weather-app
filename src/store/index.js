import Vue from "vue";
import Vuex from "vuex";
import axios from "axios";

Vue.use(Vuex);

const data = {
  coord: { lon: -0.13, lat: 51.51 },
  weather: [
    {
      id: 300,
      main: "Drizzle",
      description: "light intensity drizzle",
      icon: "09d"
    }
  ],
  base: "stations",
  main: {
    temp: 28,
    pressure: 1012,
    humidity: 81,
    temp_min: 27,
    temp_max: 28
  },
  visibility: 10000,
  wind: { speed: 4.1, deg: 80 },
  clouds: { all: 90 },
  dt: 1485789600,
  sys: {
    type: 1,
    id: 5091,
    message: 0.0103,
    country: "GB",
    sunrise: 1485762037,
    sunset: 1485794875
  },
  id: 2643743,
  name: "London",
  cod: 200
};

export default new Vuex.Store({
  state: {
    data: data,
    apiKey: "ef6fded646b25aeb80ae34d522bb493d",
    baseUrl: "https://api.openweathermap.org/data/2.5/weather",
    isLoading: false
  },
  mutations: {
    setData(state, data) {
      state.data = data;
    }
  },
  actions: {
    fetchData({ commit, state }, params) {
      state.isLoading = true;

      axios
        .get(`${state.baseUrl}${params}&appid=${state.apiKey}&units=metric`)
        .then(({ data }) => {
          commit("setData", data);
        })
        .finally(() => {
          state.isLoading = false;
        });
    }
  }
});
