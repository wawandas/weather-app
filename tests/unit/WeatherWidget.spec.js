import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";

import axios from "axios";
import MockAdapter from "axios-mock-adapter";

import { mount, createLocalVue } from "@vue/test-utils";
import WeatherWidget from "@/components/WeatherWidget.vue";
import store from "@/store";

Vue.filter("celsius", value => {
  return `${Math.round(value)}\xB0C`;
});

Vue.filter("formatedDate", timestamp => {
  const d = new Date(timestamp * 1000);
  const weekday = new Array("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat");
  const day = weekday[d.getDay()];
  const date = d.getDate();
  const hour = d.getHours();
  const min = d.getMinutes();
  return `${date} ${day}, ${hour}:${min}`;
});

Vue.use(Vuetify);

const data = {
  coord: { lon: -0.13, lat: 51.51 },
  weather: [
    {
      id: 300,
      main: "Drizzle",
      description: "light intensity drizzle",
      icon: "09d"
    }
  ],
  base: "stations",
  main: {
    temp: 28,
    pressure: 1012,
    humidity: 81,
    temp_min: 27,
    temp_max: 28
  },
  visibility: 10000,
  wind: { speed: 4.1, deg: 80 },
  clouds: { all: 90 },
  dt: 1485789600,
  sys: {
    type: 1,
    id: 5091,
    message: 0.0103,
    country: "GB",
    sunrise: 1485762037,
    sunset: 1485794875
  },
  id: 2643743,
  name: "London",
  cod: 200
};

describe("WeatherWidget.vue", () => {
  let wrapper;

  beforeEach(() => {
    const localVue = createLocalVue();

    localVue.use(Vuetify);
    localVue.use(Vuex);

    wrapper = mount(WeatherWidget, {
      store,
      localVue
    });
  });

  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("fetches weather data of entered city on enter event", async () => {
    const mock = new MockAdapter(axios);

    mock.onGet(/\/weather/).reply(200, {
      ...data,
      sys: {
        ...data.sys,
        country: "DE"
      },
      name: "Berlin"
    });

    wrapper.find('input[name="city"]').setValue("Berlin");
    wrapper.find('input[name="city"]').trigger("keydown.enter");

    //this part has to be improved :'((
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();

    expect(wrapper.find('div[data-label="weather-city-name"]').text()).toBe(
      "Berlin, DE"
    );
  });
});
