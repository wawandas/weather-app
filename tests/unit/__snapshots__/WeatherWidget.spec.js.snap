// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherWidget.vue renders correctly 1`] = `
<div class="mx-auto v-card v-sheet theme--light" style="max-width: 400px;">
  <form novalidate="novalidate" class="v-form pa-4">
    <div class="row">
      <div class="col col-10">
        <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--enclosed v-text-field--placeholder">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><input name="city" id="input-6" placeholder="Enter city" type="text"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--clear"><button disabled="disabled" type="button" aria-label="clear icon" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">$clear</button></div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="align-center col col-2"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-map-marker theme--light"></i></span></button></div>
    </div>
  </form>
  <div>
    <div tabindex="-1" class="v-list-item v-list-item--two-line theme--light">
      <div class="v-list-item__content">
        <div data-label="weather-city-name" class="v-list-item__title headline">London</div>
        <div class="v-list-item__subtitle">
          30 Mon, 16:20, Drizzle
        </div>
      </div>
    </div>
    <div class="v-card__text">
      <div class="row align-center">
        <div class="col col-6"><span class="display-3">28Â°C</span></div>
        <div class="justify-center col col-6"><img src=""> <br>
          (light intensity drizzle)
        </div>
      </div>
    </div>
    <div tabindex="-1" class="v-list-item theme--light">
      <div class="v-list-item__icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-send theme--light"></i></div>
      <div class="v-list-item__subtitle">4.1 m/c</div>
    </div>
    <div tabindex="-1" class="v-list-item theme--light">
      <div class="v-list-item__icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-cloud-download theme--light"></i></div>
      <div class="v-list-item__subtitle">81%</div>
    </div>
    <div class="v-card__actions"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-cached theme--light"></i></span></button></div>
  </div>
</div>
`;
